File: filesystem.go
package main

import (
	"io/fs"
	"os"
	"path/filepath"
)

// readDir reads the contents of a directory and returns a slice of fs.DirEntry,
// filtering out special directories like .git, .gemini, and hidden files.
func readDir(dirPath string) ([]fs.DirEntry, error) {
	entries, err := os.ReadDir(dirPath)
	if err != nil {
		return nil, err
	}

	var files []fs.DirEntry
	for _, entry := range entries {
		// Ignore special directories and hidden files
		if entry.IsDir() && (entry.Name() == ".git" || entry.Name() == ".gemini") {
			continue
		}
		if !entry.IsDir() && filepath.Base(entry.Name()) == ".DS_Store" {
			continue
		}
		// Ignore hidden files and directories (those starting with a dot)
		if len(entry.Name()) > 0 && entry.Name()[0] == '.' && entry.Name() != "." && entry.Name() != ".." {
			continue
		}
		files = append(files, entry)
	}
	return files, nil
}

// readFileContent reads the entire content of a file and returns it as a string.
// It handles potential errors during file reading.
func readFileContent(filePath string) (string, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

File: python/filesystem.py
import os


def read_dir(dir_path):
    """Read directory entries, filtering hidden files, .git, .gemini, .DS_Store."""
    entries = sorted(os.scandir(dir_path), key=lambda e: e.name)
    result = []
    for entry in entries:
        name = entry.name
        # Ignore .git and .gemini directories
        if entry.is_dir() and name in (".git", ".gemini"):
            continue
        # Ignore .DS_Store files
        if not entry.is_dir() and name == ".DS_Store":
            continue
        # Ignore hidden files/directories (starting with dot)
        if name.startswith(".") and name not in (".", ".."):
            continue
        result.append(entry)
    return result


def read_file_content(file_path):
    """Read the entire content of a file and return it as a string."""
    with open(file_path, "r") as f:
        return f.read()
